### ISSUE TITLE
[Feature Request] Provide a way to call Mojo functions from Python

### ISSUE THREAD

**Issue by shashankprasanna** [modularml/mojo#1210]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

Mojo Python interop supports calling Python functions directly from Mojo. This is great for developers starting their projects in Mojo and leveraging the Python ecosystem to fill in the gaps for specific capabilities.

However, Python developers who want to accelerate compute intensive sections of their applications in Mojo don't have a clean or native way of doing this. 
This has been requested previously: 
https://stackoverflow.com/questions/77063677/how-to-pass-mojo-function-to-python-in-python-interop
And as a doc clarification request:
https://github.com/modularml/mojo/issues/621

### What is your motivation for this change?

Reducing the barrier to leveraging Mojo's performance and capabilities without completely migrating projects.

**Comment by CloseChoice** [comment]
This would be amazing. I am contributing to shap and and would like to move our C code to mojo in the future. The problem is that I would have to move the whole library to mojo which would basically be the end of it, since everybody else is using Python.

**Comment by guidorice** [comment]
🎉  this would be an awesome feature!

**Comment by mrmattwright** [comment]
Yes this would be absolutely terrific, We have couple of projects that have performance hot spots that swapping in a call to mojo would make an enormous difference, but the code base is large enough to mean a wholesale change to mojo would be a lengthy endeavour. 

Also unrelated shout out @CloseChoice - very happy user of Shap! 🙏

**Comment by morne** [comment]
I would like this feature as well.

**Comment by sa-** [comment]
This is key to getting any adoption within the python community

**Comment by haifeng-jin** [comment]
This would be a killer feature! But would like to see what are the challenges of supporting this.

**Comment by Prodesire** [comment]
This is definitely a good idea! Although Python can be accelerated through methods such as using C and Rust, the languages are quite different and it is not easy to use. If this can be achieved, the syntax is very similar, and the development and operation efficiency can be greatly improved.

**Comment by MamadouSDiallo** [comment]
This is the feature that I am waiting for to jump on the mojo bandwagon. Can't wait to see this happen.

**Comment by Moosems** [comment]
Would be very useful when passing mojo functions as a callback

**Comment by miili** [comment]
This would be great! Some code bases are just too heavy to be moved completely to mojo, also those libraries are often utilized heavily downstream.

I would love to use Mojo to replace some heavy lifting CPython functions. First PoCs show a 3-4x speed up.

**Comment by miili** [comment]
@gabrieldemarmiesse, is there an update on this language feature?

**Comment by MamadouSDiallo** [comment]
@gabrieldemarmiesse During the July 15th community meeting, they announced that it is the next big feature and it will start happening within the next three months. (https://www.youtube.com/watch?v=_QVs626Vn2k)

**Comment by MamadouSDiallo** [comment]
Any update on this issue? Happy new year.

**Comment by BradLarson** [comment]
Sorry for forgetting to revisit this older issue, but earlier this year we did indeed add this functionality to Mojo. For more, see our documentation on Python interoperability both to and from Mojo: https://docs.modular.com/mojo/manual/python/ as well as examples of this functionality in action: https://github.com/modular/modular/tree/main/examples/mojo/python-interop

Thanks for all the interest, everyone!
